catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserver.java

```diff
@@ -0,0 +1,311 @@
+
+/**
+ * Based on {@link org.apache.commons.io.monitor.FileAlterationObserver}, except modified to only
+ * update the observer state on successful async request
+ *
+ * <p>This implementation only works with one AsyncFileAlterationListener.
+ *
+ * <p>Every time the AsyncFileAlterationObserver is polled by calling {@code checkAndNotify()}, the
+ * observer will check all the {@link File}'s and {@link AsyncFileEntry}'s that are under the
+ * directory being monitored, and call the {@link AsyncFileAlterationListener}'s corresponding
+ * methods
+ *
+ * @see AsyncFileAlterationListener
+ */
+public class AsyncFileAlterationObserver implements Serializable {
+
+  private final AsyncFileEntry rootFile;
+  private AsyncFileAlterationListener listener = null;
+  private final transient Set<AsyncFileEntry> processing = new ConcurrentSkipListSet<>();
+  private static final Logger LOGGER = LoggerFactory.getLogger(AsyncFileAlterationObserver.class);
+
+  public AsyncFileAlterationObserver(File fileToObserve) {
+    rootFile = new AsyncFileEntry(fileToObserve);
+  }
+
+  @VisibleForTesting
+  protected AsyncFileEntry getRootFile() {
+    return rootFile;
+  }
+
+  public void addListener(final AsyncFileAlterationListener listener) {
+    if (this.listener != null) {
+      throw new IllegalStateException("Cannot be two listeners");
+    }
+    this.listener = listener;
+  }
+
+  public void removeListener() {
+    //  To be truly thread safe, this should block until processes is empty.
+    if (listener != null) {
+      this.listener = null;
+    } else {
+      throw new IllegalStateException("Cannot remove a listener from an empty list");
+    }
+  }
+
+  /**
+   * Fire directory/file created events to the registered listeners.
+   *
+   * @param entry The file entry
+   */
+  private void doCreate(AsyncFileEntry entry) {
+
+    if (!addToProcessors(entry)) {
+      return;
+    }
+
+    AsyncFileEntry temp = entry.getFromParent();
+
+    if (temp != null) {
+      //  Someone already committed us. so we're done.
+      removeFromProcessors(entry);
+      return;
+    }
+
+    if (!entry.getFile().isDirectory()) {
+
+      LOGGER.trace("Sending create Request for {}", entry.getName());
+
+      listener.onFileCreate(
+          entry.getFile(), new CompletionSynchronization(entry, this::commitCreate));
+    } else {
+      // Directories are always committed and added to the parent IF they
+      // don't already exist
+
+      File[] children = listFiles(entry.getFile());
+      for (File child : children) {
+        doCreate(new AsyncFileEntry(entry, child));
+      }
+
+      commitCreate(entry, true);
+    }
+  }
+
+  /**
+   * Callback to allow successful ContentFiles to commit themselves
+   *
+   * @param entry The AsyncFileEntry wrapping the file being listened to.
+   * @param success Boolean that shows if the task failed or completed successfully
+   */
+  private void commitCreate(AsyncFileEntry entry, boolean success) {
```

Comment:
It's also used as a callback function in the CompletionSynchronization where it can be false.

Alternatively this could be refactored to take an on success and on failure callback if that would look cleaner.